src/crxviewer.js

```diff
@@ -887,6 +895,73 @@ var viewFileInfo = (function() {
     return viewFileInfo;
 })();
 
+function parseExtensionFromManifest(manifest) {
+    var id;
+    if (manifest.browser_specific_settings &&
+        manifest.browser_specific_settings.gecko &&
+        (id = manifest.browser_specific_settings.gecko.id)) {
+        return id;
+    }
+
+    if (manifest.applications && manifest.applications.gecko &&
+        (id = manifest.applications.gecko.id)) {
+        return id;
+    }
+
+    return '';
+}
+
+// Extract the extension ID from the currently loaded extension. The ID will be
+// determined from the COSE signature, the PKCS#7 signature, or manifest.json,
+// whatever comes first. Resolves to the extension ID (and the extension type
+// between parentheses if available), or rejects on parser errors.
+function calculateExtensionID() {
+    var fileList = document.getElementById('file-list');
+    var readEntry = function(filename, Writer) {
+        var li = fileList.querySelector('li[data-filename="' + filename + '"]');
+        if (!li) {
+            return null;
+        }
+        return new Promise(function(resolve) {
+            li.zipEntry.getData(new Writer(), function(data) {
+                resolve(data);
+            });
+        });
+    };
+    var formatSubject = function(subject) {
+        return subject.CN + ' (' + subject.OU + ')';
+    };
+
+    return Promise.all([
+        readEntry('manifest.json', EfficientTextWriter),
+        readEntry('META-INF/cose.sig', Uint8ArrayWriter),
+        readEntry('META-INF/mozilla.rsa', Uint8ArrayWriter),
+    ]).then(function(entries) {
+        var manifest_json = entries[0];
+        var cose_sig = entries[1];
+        var pkcs7_sig = entries[2];
+
+        if (cose_sig) {
+            return formatSubject(parseCertificate(parseMozCOSE(cose_sig)));
+        }
+
+        if (pkcs7_sig) {
+            return formatSubject(parseCertificate(pkcs7_sig));
+        }
+
+        // manifest.json is usually checked first, but it does not contain the
+        // certification source (e.g. Production), so check it after others.
+        if (manifest_json) {
+            var id = parseExtensionFromManifest(JSON.parse(manifest_json));
```

Comment:
`JSON.parse` does not necessarily work, because `manifest.json` may contain comments.

Use `JSON.minify` from https://github.com/Rob--W/crxviewer/blob/f556e326bab7a1dd9839221ebb66183cd0776f65/src/lib/beautify/minify.json.js
e.g. as seen at https://github.com/Rob--W/crxviewer/blob/f556e326bab7a1dd9839221ebb66183cd0776f65/src/lib/beautify/beautify-json.js#L12-L17

Reply:
Ack, documented at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json